orm/src/main/java/org/hibernate/search/cfg/impl/HibernateCoreIdUniquenessResolver.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+
+package org.hibernate.search.cfg.impl;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+
+import org.hibernate.cfg.Configuration;
+import org.hibernate.mapping.PersistentClass;
+import org.hibernate.search.cfg.spi.IdUniquenessResolver;
+
+/**
+ * Implementation extracting entity classes from the configuration
+ * and release the configuration reference for future Garbage Collection.
+ *
+ * @author Emmanuel Bernard &lt;emmanuel@hibernate.org&gt;
+ */
+public class HibernateCoreIdUniquenessResolver implements IdUniquenessResolver {
+	private final Set<Class<?>> entities;
+
+	public HibernateCoreIdUniquenessResolver(Configuration cfg) {
+		Set<Class<?>> entities = new HashSet<>();
+		Iterator<PersistentClass> iterator = cfg.getClassMappings();
+		while ( iterator.hasNext() ) {
+			entities.add( iterator.next().getMappedClass() );
+		}
+		this.entities = Collections.unmodifiableSet( entities );
+	}
+
+	@Override
+	public boolean areIdsUniqueForClasses(Class<?> entityInIndex, Class<?> otherEntityInIndex) {
+		/*
+		 * Look for the top most superclass of each that is also a mapped entity
+		 * That should be the root entity for that given class.
+		 */
+		Class<?> rootOfEntityInIndex = getRootEntity( entityInIndex );
+		Class<?> rootOfOtherEntityInIndex = getRootEntity( otherEntityInIndex );
+		return rootOfEntityInIndex == rootOfOtherEntityInIndex;
+	}
+
+	private Class<?> getRootEntity(Class<?> entityInIndex) {
+		if ( ! entities.contains( entityInIndex ) ) {
+			// the entity is not a mapped entity ?
```

Comment:
I think I prefer to return the entity itself rather than this exception. It is not inherently invalid for Hibernate Search to index non JPA entities.

We could add a log in the system to say whether or not we had to disable that optimization though when we build the visitor. That's probably a good idea. A user will spot the safe but slow mode and will be able to investigate.